mitiq/mitiq_qiskit/conversions.py

```diff
@@ -41,6 +46,155 @@ def _remove_barriers(circuit: QuantumCircuit) -> QuantumCircuit:
     return copy
 
 
+def _map_bit_index(
+    bit_index: int, new_register_sizes: List[int]
+) -> Tuple[int, int]:
+    """Returns the register index and (qu)bit index in this register for the
+    mapped bit_index.
+
+    Args:
+        bit_index: Index of (qu)bit in the original register.
+        new_register_sizes: List of sizes of the new registers.
+
+    Example:
+        bit_index = 3, new_register_sizes = [2, 3]
+        returns (1, 0), meaning the mapped (qu)bit is in the 1st new register
+        and has index 0 in this register.
+
+    Note:
+        The bit_index is assumed to come from a circuit with 1 or n registers
+        where n is the maximum bit_index.
+    """
+    max_indices_in_registers = np.cumsum(new_register_sizes) - 1
+
+    # Could be faster via bisection.
+    register_index = None
+    for i in range(len(max_indices_in_registers)):
+        if bit_index <= max_indices_in_registers[i]:
+            register_index = i
+            break
+    assert register_index is not None
+
+    if register_index == 0:
+        return register_index, bit_index
+
+    return (
+        register_index,
+        bit_index - max_indices_in_registers[register_index - 1] - 1,
+    )
+
+
+def _map_bits(bits, registers, new_register_sizes, new_registers):
+    """Maps (qu)bits to new registers. Assumes the input ``bits`` come from
+    a single register or n registers, where n is the number of bits.
+    """
+    if len(new_registers) == 0:
+        return bits
+
+    # Only two support cases:
+    if len(registers) == 1:
+        # Case where there are n bits in a single register.
+        indices = [bit.index for bit in bits]
+    else:
+        # Case where there are n single-bit registers.
+        indices = [registers.index(bit.register) for bit in bits]
+
+    mapped_indices = [_map_bit_index(i, new_register_sizes) for i in indices]
+
+    if isinstance(new_registers[0], qiskit.QuantumRegister):
+        Bit = qiskit.circuit.Qubit
+    else:
+        Bit = qiskit.circuit.Clbit
+
+    return [Bit(new_registers[i], j) for i, j in mapped_indices]
+
+
+def _transform_registers(
```

Comment:
I am surprised this function doesn't exist in qiskit. But indeed I could not find it. 

However, I found something similar which allows to append an "old" circuit to a new circuit.
It can be done as follows:
`new_circuit.append(c.to_instruction(), new_circuit.qubits).decompose()`

In this way, one could assign a circuit with old registers to a circuit with new registers without the need to
define qubit mapping functions.

This is an optional suggestion to simplify the code but the current solution is fine for me as long as it works.

Reply:
Ah, thanks for the suggestion. Let me look into this, I prefer to use code that already exists inasmuch possible.

Reply:
After wrestling with Qiskit for the afternoon, I think I want to retract my previous comment.

This strategy is very brittle with a new error popping up for each unit test. Probably the simplest is when you have an unused classical register:

```python
import qiskit

old = qiskit.QuantumCircuit(qiskit.QuantumRegister(1), qiskit.ClassicalRegister(1))
old.h(old.qregs[0])
print(old)

# Try to transform registers.
new = qiskit.QuantumCircuit(qiskit.QuantumRegister(1), qiskit.ClassicalRegister(1))
new.append(old.to_instruction(), qargs=new.qubits, cargs=new.clbits)
print(new.decompose())
```

which raises this error:

```python
DAGCircuitError: 'Cannot replace node of width (1 qubits, 1 clbits) with instruction of mismatched width (1 qubits, 0 clbits).'
```

This can be fixed but it requires performing checks on the instruction, and then we're essentially back to the current implementation.

There's a bunch of other errors for other unit tests, even after fixing a bunch. I'm ready to scrap this strategy unless someone else wants to take a stab at it.